activesupport/lib/active_support/core_ext/enumerable.rb

```diff
@@ -16,6 +16,22 @@ module Enumerable
 
   # :startdoc:
 
+  # Calculates the minimum from the extracted elements.
+  #
+  # payments = [Payment.new(5), Payment.new(15), Payment.new(10)]
+  # payments.minimum(:price) # => 5
+  def minimum(key)
+    pluck(key).min
+  end
+
+  # Calculates the maximum from the extracted elements.
+  #
+  # payments = [Payment.new(5), Payment.new(15), Payment.new(10)]
+  # payments.maximum(:price) # => 15
+  def maximum(key)
+    pluck(key).max
```

Comment:
`pluck` extracts the `key` via `#[]`.  Is that what we want here?  Or should we treat `key` as a method name?  e.g.:

```suggestion
    max_by(&key).send(key)
```

Reply:
Good question. I think it depends what behavior we want here.

In case of the `fresh_when` example it'd be nice to fallback to the existing logic (returning `nil`). That being said it might be an edge case that an array is passed that does not respond to `updated_at`.

If not, would the following be too crazy:

```rb
payments.maximum(:processed_at) # => nil
payments.maximum(&:processed_at) # => NoMethodError (undefined method `processed_at' for #<Payment id: 135138680>)
```

Reply:
> In case of the `fresh_when` example it'd be nice to fallback to the existing logic (returning `nil`). That being said it might be an edge case that an array is passed that does not respond to `updated_at`.

I'm not sure I follow.  As long as the array elements respond to `:updated_at`, the `fresh_when` use-case should be the same, no?  Or are you saying that you would like to be able pass an array to `fresh_when` wherein each element does not respond to `:updated_at` (but does respond to `:[]`)?

But speaking of `nil`, there is a corner case that I think we should add a test for: an empty enumerable.  The suggestion I made above could be modified as `max_by(&key)&.send(key)` to handle an empty enumerable, but it turns out that `max_by` is surprisingly slow!:

<details>
  <summary><code>max_by</code> benchmark</summary>

  ```ruby
  require "benchmark/ips"

  VALUES = (1..20).map { |n| "x" * n }
  KEY = :length

  Benchmark.ips do |job|
    job.report("map + max") do
      VALUES.map(&KEY).max
    end

    job.report("just max_by") do
      VALUES.max_by(&KEY) # without `send(KEY)`
    end

    job.compare!
  end
  ```
</details>

```
Comparison:
           map + max:   228288.2 i/s
         just max_by:   188118.4 i/s - 1.21x  (± 0.00) slower
```

So it might be better to use `map` + `max`:

```suggestion
    map(&key).max
```

Reply:
> I'm not sure I follow. As long as the array elements respond to `:updated_at`, the `fresh_when` use-case should be the same, no? 

Yep it will be, was just thinking out loud in case the array elements would not respond to `:updated_at` eg. when records do not have the `timestamp` columns. 

With `timestamp` columns (the happy path): 

```rb
payments = [Payment.new]
payments.pluck(:updated_at).max # => Tue, 09 Feb 2021 10:20:19.238011000 UTC +00:00
payments.map(&:updated_at).max  # => Tue, 09 Feb 2021 10:20:19.238011000 UTC +00:00
```

Without timestamp columns (the edge case):

```rb
payments = [Payment.new]
payments.pluck(:updated_at).max # => nil
payments.map(&:updated_at).max # => NoMethodError (undefined method `updated_at' for #<Payment id: 135138680>)
```

I don't have a strong preference here, might be okay to assume the `timestamp` columns will be available, but happy to leave the decision to the experts.  

> [...] but it turns out that `max_by` is surprisingly slow! [...] So it might be better to use `map` + `max`

👍

Reply:
> Yep it will be, was just thinking out loud in case the array elements would not respond to `:updated_at` eg. when records do not have the `timestamp` columns.

In such a case, passing a relation (e.g. `Payment.all`) to `fresh_when` would also raise, so I don't think we need to support it.

Reply:
That's super helpful context, thank you

Reply:
Pushed the second suggestion. Thanks @jonathanhefner